# Map, Filter, Reduce, Lambda & Recursion

## Tasks Today:

1) <b>Lambda Functions</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Saving to a Variable <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Multiple Inputs <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Passing a Lambda into a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; e) Returning a Lambda from a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; f) In-Class Exercise #1 <br>
2) <b>Map</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Map <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #2 <br>
3) <b>Filter</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Filter <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #3 <br>
4) <b>Reduce</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Reduce <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #4 <br>
5) <b>Recursion</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Implementing a Base <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Writing a Factorial Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #5 <br>
6) <b>Generators & Iterators</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Yield Keyword <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Inifinite Generator <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #6 <br>
7) <b>Exercises</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Exercise #1 - Filtering Empty Strings <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Exercise #2 - Sorting with Last Name <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Exercise #3 - Conversion to Farhenheit <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Exercise #4 - Fibonacci Sequence <br>

## Lambda Functions <br>
<p>Lambda functions... or "Anonymous Functions" are referring to inline functions with no name. The keyword lambda denotes the no name function, and executes within a single line. Without saving it to a variable; however, it is not able to be used, unless passed in either as a paramater or within list comprehension.<br>Written as "(keyword lambda) (one or more inputs) (colon) (function to be executed)"</p>

#### Syntax

In [11]:
def addTwo(x):
    return(x+2)
print(addTwo(6))

# lanmbda version
# without using a variable to pass in data
# lambda is just a one time use function here were setting the input as x
# so the number 6 we typed in as our input will be used where x is in our lambda function  
print((lambda x: x + 2)(6))

#using a variable
a_num= 4
print((lambda x: x +2)(a_num))

8
8
6


#### Saving to a Variable

In [17]:
f_test= lambda x: x+3

f_test(4)

7

#### Multiple Inputs

In [20]:
# Multiple inputs with no variable
# create a one time function with lambda, then were telling it we want three inputs that we named x,y and z
# then were saying the function will multiple x * y *z
#then all we have to do is give our three inputs in the next set of braces and it will pass those numbers in as x y and z
print((lambda x, y, z: x * y* z)(3,5,8))

# multiple inputs with variable attached

x_test= lambda x,y,z: x * y * z  
print(x_test(3,5,8))


120
120


#### Passing a Lambda into a Function

In [19]:
# here we crate a function called multiple that take two inputs
# f will be a lambda function 
# num will be a number 

# were returning the lambda function(lambda x: x*x)
# then were were returning a number (num) to be used as the input x in our lambda function 
# we dont have to have the num in its own braces becasue the multiply function is putting it in its braces for us 

def multiply(f,num):
    """
    f expects a lambda funcition 
    num expects a number 
    """

    return f(num)

multiply(lambda x: x * x, 4)






16

#### Returning a Lambda from a Function

In [96]:
# # regular defined function 
# def multiply_tst(num):
#     return num * 4

# #function within a function 
# def returnFunc():
#     test=4
#     def multiply(num):
#         return num * 2
#     return multiply


# being that this function has two inputs one is () and one is (num) if we want to return the info from num we have to 
# write in two () in our print statement otherwise well get an error 
# print(returnFunc()(9))


#lambda function returned from regular function 
def returnLam(b,c):
    return lambda x, a: x + a + b + c 

r_lamb=returnLam(4,5) #here were sending our first two inputs into the returnLam(4,5) function
print(r_lamb(5,5)) # here were sending our next two arguments for out lambda funtion (x,a) because we havent given those values yet


19


#### If Statements within Lambdas

In [101]:
# here we need 1 input that we've named num
# and our function is saying multiply the number we put in the input(num) times 2 if the number is greater than 10
# in the number we put in in less than 10 just add 2 
#Lambda x: True if (condition) else False 
f_condition= lambda num : num * 2 if num > 10 else num + 2

print(f_condition(11))
print(f_condition(3))

22
5


#### In-Class Exercise #1 <br>
<p>Write an anonymous function that cubes the arguments passed in and assign the anonymous function to a variable 'f'.</p>

In [109]:
f = lambda x: x * x * x 

print(f(7))



343


## Map <br>
<p>The map function allows you to iterate over an entire list while running a function on each item of the list. This is why the map function works well with lambda's, because it simplifies things and you write less lines of code.<br>The syntax for a map function is "map(function to be used, list to be used)"<br>However, you must be careful, as the map function returns a map object, not a list. To turn it into a list we use the list() type conversion.</p>

#### Syntax

In [11]:
# map(func,iterable)
#an iterable has to be something like a list,dict,tuple,set
#normally the usage of map happens with a pre-defined function, but we can use lambdas as well 
# so a map function will take in a function and then its inputs for the function here were using square and then passing 
# in the two inputs that our squared function requires 

def squared(num,num2):
    if num <10 and num2 <10:
        return num **2, num2 **2 
    else:
        return num, num2 

numbers= [4,11,20,3,15,20]
more_nums=[4,10,3,2,6]

squared_nums_map = list(map(squared,numbers,more_nums))
print(squared_nums_map)



[(16, 16), (11, 10), (20, 3), (9, 4), (15, 6)]


#### Using Lambda's with Map

In [12]:
# map(lambda x: do something,iterable)
# using lambda in map happens inline or on onr line 
#lists(map(lamba x; x**2,nums))

squared_nums_lambda= list(map(lambda x,y:(x**2, y**2)if x <10 and y<10 else(x,y), numbers, more_nums))

print (squared_nums_lambda)


[(16, 16), (11, 10), (20, 3), (9, 4), (15, 6)]


#### In-Class Exercise #2 <br>
<p>Use the map function to double each number and minus it by one in the list by using a lambda function</p>

In [16]:
jons_lambda= list(map(lambda a,b: (a*2-1, b*2-1),numbers,more_nums))

print(jons_lambda)

[(7, 7), (21, 19), (39, 5), (5, 3), (29, 11)]


## Filter() <br>
<p>Filter's are similar to the map function, where you're able to pass a function argument and a list argument and filter out something from the list based on the conditions passed. Similar to the map function, it returns a filter object, so you need to type convert it to a list()</p>

#### Syntax

In [230]:
#here we are creating a function that is going to look for an item that has the letter a as the first letter in an item and 
#will return it as true if the item has the letter a 
#then we made a new variable called New_names_list we can then use the filter function to only print the names that return as true

names= ["Bob","Andy","Max","Evan","Angelica"]

def a_names(name):
    if name[0].lower()==" ":
        return True
    else: 
        return False
    
new_names_list = list(filter(a_names,names))
print(new_names_list)


[' df ']


#### Using Lambda's with Filter()

In [179]:
new_names_lambda= list(filter(lambda name: True if name[0].lower()=="a" else False,names))
print(new_names_lambda)

['Andy', 'Angelica']


#### In-Class Exercise #3 <br>
<p>Filter out all the numbers that are below the mean of the list.<br><b>Hint: Import the 'statistics' module</b></p>

In [45]:
from statistics import mean

numbers = [4, 11, 20, 3, 15, 20]

jon_filter= list(filter(lambda number: True if number > mean(numbers) else False, numbers))

print(jon_filter)



[20, 15, 20]


## Reduce() <br>
<p>Be very careful when using this function, as of Python 3 it's been moved to the 'functools' library and no longer is a built-in function.<br>The creator of Python himself, says to just use a for loop instead.</p>

#### Syntax

In [50]:
# here we are taking each item in the the list and adding them to eachother 
#this i a function thats no longr used but we may see it in the wild in older projects 
# this will do the same thing for subtraction and multiplication where we will just 
#keep multiplying one number to the next iin our list 

from functools import reduce

#reduce(function,iterable)

list_1= [2,4,6,8,10]

def addNums(num1,num2):
    return num1+num2

result_add= reduce(addNums, list_1)

print(result_add)

30


#### Using Lambda's with Reduce()

In [51]:
result_lamb= reduce(lambda x,y: x+y,list_1)
print(result_lamb)

30


#### In-Class Exercise #4 <br>
<p>Use the reduce function to multiply the numbers in the list below together with a lambda function.</p>

In [64]:
my_list= [1,2,3,4]

jon_rd= reduce(lambda x,y: x*y,my_list)
print(jon_rd)

24


## Recursion <br>
<p>Recursion means that a function is calling itself, so it contanstly executes until a base case is reached. It will then push the returning values back up the chain until the function is complete. A prime example of recursion is computing factorials... such that 5! (factorial) is 5*4*3*2*1 which equals 120.</p>

#### Implementing a Base Case

In [96]:
def addNums(num):
    #here is our base case
    if num <=1:
        print("addNums(1) = 1")
        return num
    else:
        print(f"addNums({num})={num}+addNums({num-1})")
        return addNums(num-1)

addNums(5)




addNums(5)=5+addNums(4)
addNums(4)=4+addNums(3)
addNums(3)=3+addNums(2)
addNums(2)=2+addNums(1)
addNums(1) = 1


1

#### Writing a Factorial Function

In [144]:
# 5! = 5*4*3*2*1
# here we are passing in a number and telling the the function to do the calculation in our 
#else statement and then go back to the beginning of the function with our number we
# recall the 
def factorial(num):
    if num <= 1:
        return 1
    else:
        return num * factorial(num-1)
    
factorial(5)

120

#### In-Class Exercise #5 <br>
<p>Write a recursive function that subtracts all numbers to the argument given.</p>

In [139]:
#return num -1 subNums(num-1) 
#subNums(10)


def jon2(num):
    if num<=1:
        return num
    else:
        return num - jon2(num-1)



jon2(10)

5

## Generators <br>
<p>Generators are a type of iterable, like lists or tuples. They do not allow indexing, but they can still be iterated through with for loops. They are created using functions and the yield statement.</p>

#### Yield Keyword <br>
<p>The yield keyword denotes a generator, it doesn't return so it won't leave the function and reset all variables in the function scope, instead it yields the number back to the caller.</p>

In [157]:
def my_range(start,stop,step=2):
    while start < stop:
        yield start # yield keyword means its a generator #this is similar to return where it will send to our output but it remembers the data each time theres a change, for example when we get to the end of this loop which is the start(2)+=step(2) the next start number is now 4. when we get to the yeild again it will output 4 as well as remember that the current start number is 4 for everything below it
        start+= step

for i in my_range(2,20):
    my_generator_value= i
    print(my_generator_value)


2
4
6
8
10
12
14
16
18


#### Infinite Generator

In [ ]:
# bad, never create infinite loops
#be sure to have something that will stop your generator 

#### In-Class Exercise #6 <br>
<p>Create a generator that takes a number argument and yields that number squared, then prints each number squared until zero is reached.</p>

In [163]:
def jons_gen(num):
    while num > 0:
        yield num * num 
        num -= 1

for i in jons_gen(5):
    print(i)

25
16
9
4
1


# Exercises

### Exercise #1 <br>
<p>Filter out all of the empty strings from the list below</p>

`Output: ['Argentina', 'San Diego', 'Boston', 'New York']`

In [270]:
places = [" ","Argentina", " ", "San Diego","","  ","","Boston","New York"]

def remove_blank_items(locations):
    return [i for i in locations if i.strip()]

remove_blank_items(places)






['Argentina', 'San Diego', 'Boston', 'New York']

### Exercise #2 <br>
<p>Write an anonymous function that sorts this list by the last name...<br><b>Hint: Use the ".sort()" method and access the key"</b></p>

`Output: ['Victor aNisimov', 'Gary A.J. Bernstein', 'Joel Carter', 'Andrew P. Garfield', 'David hassELHOFF']`

In [281]:
author = ["Joel Carter", "Victor aNisimov", "Andrew P. Garfield","David hassELHOFF","Gary A.J. Bernstein"]

author.sort(key=lambda x: x.split()[-1].capitalize())

print(author)





['Victor aNisimov', 'Gary A.J. Bernstein', 'Joel Carter', 'Andrew P. Garfield', 'David hassELHOFF']


### Exercise #3 <br>
<p>Convert the list below from Celsius to Farhenheit, using the map function with a lambda...</p>

`Output: [('Nashua', 89.6), ('Boston', 53.6), ('Los Angelos', 111.2), ('Miami', 84.2)]
`

In [337]:
# F = (9/5)*C + 32
places = [('Nashua',32),("Boston",12),("Los Angelos",44),("Miami",29)]


convert_temp = list(map(lambda x: (x[0], (9/5) * x[1] + 32), places))

print(convert_temp)









[('Nashua', 89.6), ('Boston', 53.6), ('Los Angelos', 111.2), ('Miami', 84.2)]


### Exercise #4 <br>
<p>Write a recursion function to perform the fibonacci sequence up to the number passed in.</p>

`Output for fib(5) => 
Iteration 0: 1
Iteration 1: 1
Iteration 2: 2
Iteration 3: 3
Iteration 4: 5
Iteration 5: 8`

In [397]:
def fibonacci_until_number(target):
    fibonacci_sequence = [0, 1]

    while fibonacci_sequence[-1] + fibonacci_sequence[-2] <= target:
        next_number = fibonacci_sequence[-1] + fibonacci_sequence[-2]
        fibonacci_sequence.append(next_number)

    Iteration=0
        
    for i in fibonacci_sequence:
        print(f"Iteration {Iteration}: {i}")
        Iteration += 1
        
        


fibonacci_until_number(100)








Iteration 0: 0
Iteration 1: 1
Iteration 2: 1
Iteration 3: 2
Iteration 4: 3
Iteration 5: 5
Iteration 6: 8
Iteration 7: 13
Iteration 8: 21
Iteration 9: 34
Iteration 10: 55
Iteration 11: 89
